File: frontend/src/api/alerts.ts
import apiClient from './apiClient';

export interface Alert {
  id: number;
  alert_type: string;
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  title: string;
  message: string;
  metric_value: number | null;
  created_at: string;
  acknowledged_at: string | null;
}

export interface RemediationModule {
  id: number;
  student_id: number;
  title: string;
  description: string | null;
  skill_gap: string;
  difficulty_level: string;
  content: string;
  content_type: string;
  completion_percentage: number;
  score: number | null;
  started_at: string | null;
  completed_at: string | null;
}

export interface ReflectionPrompt {
  id: number;
  prompt_text: string;
  context: string;
  generated_at: string;
  response: string | null;
  response_submitted_at: string | null;
  reflection_depth_score: number | null;
  sentiment: number | null;
}

export interface EthicalProfile {
  id: number;
  student_id: number;
  academic_integrity_score: number;
  collaboration_fairness_score: number;
  self_regulation_score: number;
  responsibility_index: number;
  integrity_flags: number;
  collaboration_violations: number;
  self_plagiarism_detected: boolean;
  last_violation_at: string | null;
  intervention_sent: boolean;
}

export interface EngagementSnapshot {
  id: number;
  student_id: number;
  engagement_score: number;
  activity_count: number;
  engagement_trend: string;
  timestamp: string;
}

// ============================================================================
// ALERT ENDPOINTS
// ============================================================================

export const fetchStudentAlerts = async (studentId: number, unacknowledgedOnly: boolean = false): Promise<Alert[]> => {
  try {
    const response = await apiClient.get(`/api/alerts/student/${studentId}`, {
      params: { unacknowledged_only: unacknowledgedOnly }
    });
    return response.data || [];
  } catch (error) {
    console.error('Error fetching alerts:', error);
    return [];
  }
};

export const acknowledgeAlert = async (alertId: number, acknowledgedById?: number, reason?: string): Promise<Alert | null> => {
  try {
    const response = await apiClient.post(`/api/alerts/${alertId}/acknowledge`, {
      acknowledged_by_id: acknowledgedById || 1,
      dismissal_reason: reason,
      action_taken: 'viewed'
    });
    return response.data;
  } catch (error) {
    console.error('Error acknowledging alert:', error);
    return null;
  }
};

export const checkAndCreateAlerts = async (studentId: number, courseId: number): Promise<any> => {
  try {
    const response = await apiClient.post(`/api/alerts/check/${studentId}/${courseId}`);
    return response.data;
  } catch (error) {
    console.error('Error checking alerts:', error);
    return null;
  }
};

export const createAlertRule = async (ruleData: any): Promise<any> => {
  try {
    const response = await apiClient.post('/api/alerts/rules', ruleData);
    return response.data;
  } catch (error) {
    console.error('Error creating alert rule:', error);
    return null;
  }
};

// ============================================================================
// REMEDIATION MODULE ENDPOINTS
// ============================================================================

export const fetchStudentModules = async (studentId: number, courseId: number): Promise<RemediationModule[]> => {
  try {
    const response = await apiClient.get(`/api/remediation/student/${studentId}/${courseId}`);
    return response.data || [];
  } catch (error) {
    console.error('Error fetching remediation modules:', error);
    return [];
  }
};

export const createRemediationModule = async (moduleData: any): Promise<RemediationModule | null> => {
  try {
    const response = await apiClient.post('/api/remediation/modules', moduleData);
    return response.data;
  } catch (error) {
    console.error('Error creating remediation module:', error);
    return null;
  }
};

export const updateModuleProgress = async (
  moduleId: number,
  completionPercentage: number,
  score?: number
): Promise<RemediationModule | null> => {
  try {
    const response = await apiClient.put(`/api/remediation/modules/${moduleId}`, {
      completion_percentage: completionPercentage,
      score
    });
    return response.data;
  } catch (error) {
    console.error('Error updating module progress:', error);
    return null;
  }
};

// ============================================================================
// REFLECTION PROMPT ENDPOINTS
// ============================================================================

export const fetchStudentPrompts = async (studentId: number): Promise<ReflectionPrompt[]> => {
  try {
    const response = await apiClient.get(`/api/reflection/student/${studentId}`);
    return response.data || [];
  } catch (error) {
    console.error('Error fetching reflection prompts:', error);
    return [];
  }
};

export const createReflectionPrompt = async (promptData: any): Promise<ReflectionPrompt | null> => {
  try {
    const response = await apiClient.post('/api/reflection/prompts', promptData);
    return response.data;
  } catch (error) {
    console.error('Error creating reflection prompt:', error);
    return null;
  }
};

export const submitReflection = async (promptId: number, response: string): Promise<ReflectionPrompt | null> => {
  try {
    const result = await apiClient.post(`/api/reflection/prompts/${promptId}/submit`, {
      response
    });
    return result.data;
  } catch (error) {
    console.error('Error submitting reflection:', error);
    return null;
  }
};

// ============================================================================
// ETHICS PROFILE ENDPOINTS
// ============================================================================

export const fetchEthicalProfile = async (studentId: number): Promise<EthicalProfile | null> => {
  try {
    const response = await apiClient.get(`/api/ethics/profile/${studentId}`);
    return response.data;
  } catch (error) {
    console.error('Error fetching ethical profile:', error);
    return null;
  }
};

export const flagIntegrityViolation = async (studentId: number, violation: any): Promise<any> => {
  try {
    const response = await apiClient.post(`/api/ethics/flag/${studentId}`, violation);
    return response.data;
  } catch (error) {
    console.error('Error flagging integrity violation:', error);
    return null;
  }
};

// ============================================================================
// ENGAGEMENT ENDPOINTS
// ============================================================================

export const fetchEngagementSnapshots = async (
  studentId: number,
  courseId: number,
  hours?: number
): Promise<EngagementSnapshot[]> => {
  try {
    const response = await apiClient.get(`/api/engagement/snapshots/${studentId}/${courseId}`, {
      params: { hours }
    });
    return response.data || [];
  } catch (error) {
    console.error('Error fetching engagement snapshots:', error);
    return [];
  }
};

export const createEngagementSnapshot = async (snapshotData: any): Promise<EngagementSnapshot | null> => {
  try {
    const response = await apiClient.post('/api/engagement/snapshot', snapshotData);
    return response.data;
  } catch (error) {
    console.error('Error creating engagement snapshot:', error);
    return null;
  }
};

export const fetchEngagementTrend = async (
  studentId: number,
  courseId: number,
  days?: number
): Promise<any> => {
  try {
    const response = await apiClient.get(`/api/engagement/trend/${studentId}/${courseId}`, {
      params: { days }
    });
    return response.data;
  } catch (error) {
    console.error('Error fetching engagement trend:', error);
    return null;
  }
};

// ============================================================================
// CURRICULUM ENDPOINTS
// ============================================================================

export const fetchLatestCurriculumSequence = async (
  studentId: number,
  courseId: number
): Promise<any> => {
  try {
    const response = await apiClient.get(`/api/curriculum/latest/${studentId}/${courseId}`);
    return response.data;
  } catch (error) {
    console.error('Error fetching curriculum sequence:', error);
    return null;
  }
};

export const applyCurriculumSequence = async (sequenceId: number): Promise<any> => {
  try {
    const response = await apiClient.post(`/api/curriculum/apply/${sequenceId}`);
    return response.data;
  } catch (error) {
    console.error('Error applying curriculum sequence:', error);
    return null;
  }
};

export const analyzeCurriculum = async (studentId: number, courseId: number, currentModules: number[]): Promise<any> => {
  try {
    const response = await apiClient.post(`/api/curriculum/analyze/${studentId}/${courseId}`, {
      current_modules: currentModules
    });
    return response.data;
  } catch (error) {
    console.error('Error analyzing curriculum:', error);
    return null;
  }
};

File: backend/app/routers/alerts.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.database import get_db
from app.services.alert_service import (
    AlertService, RemediationService, EthicsService, CurriculumService, EngagementService
)
from app.services.adaptive_sequencing import analyze_and_create_adaptive_curriculum
from app.schemas.alert import (
    AlertRuleCreate, StudentAlertResponse, StudentAlertAcknowledge,
    RemediationModuleCreate, RemediationModuleResponse, RemediationModuleUpdate,
    ReflectionPromptCreate, ReflectionPromptResponse, ReflectionPromptSubmit,
    EthicalProfileResponse, CurriculumSequenceResponse, EngagementSnapshotResponse,
    CurriculumAnalyzeRequest
)
from app.auth.dependencies import get_current_user
from typing import List

router = APIRouter(prefix="/api", tags=["alerts"])

# ============================================================================
# ALERT ENDPOINTS
# ============================================================================

@router.post("/alerts/rules", response_model=dict)
async def create_alert_rule(
    rule_data: AlertRuleCreate,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Create a new alert rule for a student"""
    rule = AlertService.create_alert_rule(db, rule_data)
    return {
        "id": rule.id,
        "alert_type": rule.alert_type,
        "threshold": rule.threshold,
        "severity": rule.severity,
        "active": rule.active
    }


@router.get("/alerts/student/{student_id}", response_model=List[StudentAlertResponse])
async def get_student_alerts(
    student_id: int,
    unacknowledged_only: bool = False,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Get alerts for a student"""
    alerts = AlertService.get_student_alerts(db, student_id, unacknowledged_only)
    return alerts


@router.post("/alerts/check/{student_id}/{course_id}", response_model=dict)
async def check_and_create_alerts(
    student_id: int,
    course_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Manually trigger alert checking"""
    alerts = AlertService.check_and_create_alerts(db, student_id, course_id)
    return {
        "alerts_created": len(alerts),
        "alerts": [
            {
                "id": a.id,
                "alert_type": a.alert_type,
                "severity": a.severity,
                "title": a.title,
                "metric_value": a.metric_value
            }
            for a in alerts
        ]
    }


@router.post("/alerts/{alert_id}/acknowledge", response_model=StudentAlertResponse)
async def acknowledge_alert(
    alert_id: int,
    ack_data: StudentAlertAcknowledge,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Acknowledge an alert"""
    alert = AlertService.acknowledge_alert(db, alert_id, ack_data)
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")
    return alert


# ============================================================================
# REMEDIATION MODULE ENDPOINTS
# ============================================================================

@router.post("/remediation/modules", response_model=RemediationModuleResponse)
async def create_remediation_module(
    module_data: RemediationModuleCreate,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Create a new remediation module"""
    module = RemediationService.create_module(db, module_data)
    return module


@router.get("/remediation/student/{student_id}/{course_id}", response_model=List[RemediationModuleResponse])
async def get_student_modules(
    student_id: int,
    course_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Get all remediation modules for a student"""
    modules = RemediationService.get_student_modules(db, student_id, course_id)
    return modules


@router.put("/remediation/modules/{module_id}", response_model=RemediationModuleResponse)
async def update_module_progress(
    module_id: int,
    update_data: RemediationModuleUpdate,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Update remediation module progress"""
    module = RemediationService.update_module_progress(
        db, module_id, 
        update_data.completion_percentage or 0, 
        update_data.score
    )
    if not module:
        raise HTTPException(status_code=404, detail="Module not found")
    return module


# ============================================================================
# REFLECTION PROMPT ENDPOINTS
# ============================================================================

@router.post("/reflection/prompts", response_model=ReflectionPromptResponse)
async def create_reflection_prompt(
    prompt_data: ReflectionPromptCreate,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Create a reflection prompt"""
    prompt = RemediationService.create_reflection_prompt(db, prompt_data)
    return prompt


@router.get("/reflection/student/{student_id}", response_model=List[ReflectionPromptResponse])
async def get_student_prompts(
    student_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Get all reflection prompts for a student"""
    prompts = RemediationService.get_student_prompts(db, student_id)
    return prompts


@router.post("/reflection/prompts/{prompt_id}/submit", response_model=ReflectionPromptResponse)
async def submit_reflection(
    prompt_id: int,
    response_data: ReflectionPromptSubmit,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Submit a reflection response"""
    prompt = RemediationService.submit_reflection(db, prompt_id, response_data)
    if not prompt:
        raise HTTPException(status_code=404, detail="Prompt not found")
    return prompt


# ============================================================================
# ETHICS PROFILE ENDPOINTS
# ============================================================================

@router.get("/ethics/profile/{student_id}", response_model=EthicalProfileResponse)
async def get_ethical_profile(
    student_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Get ethical profile for a student"""
    profile = EthicsService.get_or_create_profile(db, student_id, None)
    return profile


@router.post("/ethics/flag/{student_id}", response_model=EthicalProfileResponse)
async def flag_violation(
    student_id: int,
    violation_type: str = "SUSPICIOUS_PATTERN",
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Flag an integrity violation"""
    profile = EthicsService.flag_integrity_violation(db, student_id, violation_type)
    return profile


@router.put("/ethics/responsibility/{student_id}", response_model=EthicalProfileResponse)
async def update_responsibility(
    student_id: int,
    adjustment: float,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Update responsibility index"""
    profile = EthicsService.update_responsibility_index(db, student_id, adjustment)
    return profile


# ============================================================================
# CURRICULUM SEQUENCE ENDPOINTS
# ============================================================================

@router.post("/curriculum/sequence", response_model=CurriculumSequenceResponse)
async def create_curriculum_sequence(
    student_id: int,
    course_id: int,
    original_seq: str,
    adapted_seq: str,
    reason: str,
    skill_gaps: str,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Create curriculum sequence"""
    sequence = CurriculumService.create_sequence(
        db, student_id, course_id, original_seq, adapted_seq, reason, skill_gaps
    )
    return sequence


@router.get("/curriculum/latest/{student_id}/{course_id}", response_model=CurriculumSequenceResponse)
async def get_latest_sequence(
    student_id: int,
    course_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Get latest curriculum sequence"""
    sequence = CurriculumService.get_latest_sequence(db, student_id, course_id)
    if not sequence:
        raise HTTPException(status_code=404, detail="No sequence found")
    return sequence


@router.post("/curriculum/apply/{sequence_id}", response_model=CurriculumSequenceResponse)
async def apply_sequence(
    sequence_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Apply curriculum sequence"""
    sequence = CurriculumService.apply_sequence(db, sequence_id)
    if not sequence:
        raise HTTPException(status_code=404, detail="Sequence not found")
    return sequence


# ============================================================================
# ENGAGEMENT TRACKING ENDPOINTS
# ============================================================================

@router.post("/engagement/snapshot", response_model=EngagementSnapshotResponse)
async def create_engagement_snapshot(
    student_id: int,
    course_id: int,
    engagement_score: float,
    activity_count: int,
    avg_response_time: float,
    trend: str,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Create engagement snapshot"""
    snapshot = EngagementService.create_snapshot(
        db, student_id, course_id, engagement_score, activity_count, avg_response_time, trend
    )
    return snapshot


@router.get("/engagement/snapshots/{student_id}/{course_id}", response_model=List[EngagementSnapshotResponse])
async def get_engagement_snapshots(
    student_id: int,
    course_id: int,
    hours: int = 24,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Get recent engagement snapshots"""
    snapshots = EngagementService.get_recent_snapshots(db, student_id, course_id, hours)
    return snapshots


@router.get("/engagement/trend/{student_id}/{course_id}", response_model=dict)
async def get_engagement_trend(
    student_id: int,
    course_id: int,
    days: int = 7,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Get engagement trend"""
    trend = EngagementService.get_engagement_trend(db, student_id, course_id, days)
    return {
        "average_engagement": trend.get("average_engagement", 0.0),
        "trend": trend.get("trend", "STABLE"),
        "data_points": len(trend.get("snapshots", []))
    }


# ============================================================================
# ADAPTIVE CURRICULUM ENDPOINTS
# ============================================================================

@router.post("/curriculum/analyze/{student_id}/{course_id}", response_model=dict)
async def analyze_and_adapt_curriculum(
    student_id: int,
    course_id: int,
    req_data: CurriculumAnalyzeRequest,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Analyze student and create adaptive curriculum"""
    result = analyze_and_create_adaptive_curriculum(db, student_id, course_id, req_data.current_modules)
    return result
